manager/management/commands/export-data.py

```diff
@@ -48,77 +89,148 @@ def get_records(self):
-                                'Clicks': instance.click_count,
-                                'Recipients who clicked': instance.click_recipient_count,
-                                'Click Rate': instance.click_rate
-                            }
-
-                            writer.writerow(line)
-
-                            pbar.update(1)
+        fieldnames = [
+            'Email ID',
+            'Instance ID',
+            'Title',
+            'Subject',
+            'URL',
+            'From',
+            'Friendly',
+            'Start',
+            'End',
+            'Recipients',
+            'Sent',
+            'Opens',
+            'Open Rate',
+            'Clicks',
+            'Recipients who clicked',
+            'Click Rate'
+        ]
+
+        file_handler = None
+        email_handler = None
+
+        if self.filename:
+            file_handler = open(self.filename, 'w')
+            try:
+                file_writer = csv.DictWriter(file_handler, fieldnames=fieldnames)
+                file_writer.writeheader()
+            except:
+                file_handler.close()
+
+        if self.email:
+            email_handler = StringIO()
+            email_writer = csv.DictWriter(email_handler, fieldnames=fieldnames)
+            email_writer.writeheader()
+
+        total_instances = Instance.objects.filter(email__in=records).count()
+
+        with tqdm(total=total_instances) as pbar:
+            for record in records:
+                instances = record.instances.all()
+
+                if self.before:
+                    instances = instances.filter(end__lt=self.before)
+
+                email_title = record.title
+
+                for instance in instances.all():
+                    line = {
+                        'Email ID': record.id,
+                        'Instance ID': instance.id,
+                        'Title': email_title,
+                        'Subject': instance.subject,
+                        'URL': record.source_html_uri,
+                        'From': record.from_email_address,
+                        'Friendly': record.from_friendly_name,
+                        'Start': instance.start,
+                        'End': instance.end,
+                        'Recipients': instance.recipients.count(),
+                        'Sent': instance.sent_count,
+                        'Opens': instance.initial_opens,
+                        'Open Rate': instance.open_rate,
+                        'Clicks': instance.click_count,
+                        'Recipients who clicked': instance.click_recipient_count,
+                        'Click Rate': instance.click_rate
+                    }
+
+                    if file_handler and file_writer:
+                        try:
+                            file_writer.writerow(line)
+                        except Exception as e:
+                            file_writer.close()
```

Comment:
Should this be `file_handler.close()`, or is `file_writer.close()` correct here?

Reply:
`file_handler` is the actual file stream, which is what needs to be closed. In this case, `file_writer` is a `CSVDictWriter`, which gets a reference to a stream.

Reply:
Usually, we wouldn't do this at all... we'd use the `with open(filename, 'w') as file_handler:` syntax. However, I thought I was going to have 2 streams to deal with - which I actually don't, since StringIO handles opening/closing the IOStream automatically. But even without the 2 streams, we aren't necessarily using the file stream if `filename` isn't passed in. Since we're only conditionally creating the stream, it didn't seem to make sense to use a with statement, as it would make the code clunky.